tests/unit/miniapps/test_tmop.cpp

```diff
@@ -0,0 +1,851 @@
+      Args &JI(const double arg) { jitter = arg; return *this; }
+
+      Args &POR(set arg) { order = arg; return *this; }
+      Args &TID(set arg) { target_id = arg; return *this; }
+      Args &MID(set arg) { metric_id = arg; return *this; }
+      Args &QOR(set arg) { quad_order = arg; return *this; }
+      Args &LS(set arg) { lin_solver = arg; return *this; }
+      Args &NL(set arg) { newton_loop = arg; return *this; }
+   };
+   const char *name, *mesh;
+   int NEWTON_ITERATIONS, REFINE, LINEAR_ITERATIONS, COMBO;
+   bool NORMALIZATION;
+   double LIMITING, JITTER;
+   set P_ORDERS, TARGET_IDS, METRIC_IDS, Q_ORDERS, LINEAR_SOLVERS, NEWTON_LOOPS;
+public:
+   Launch(Args a = Args()):
+      name(a.name), mesh(a.mesh),
+      NEWTON_ITERATIONS(a.newton_iter), REFINE(a.rs_levels),
+      LINEAR_ITERATIONS(a.max_lin_iter), COMBO(a.combo),
+      NORMALIZATION(a.normalization), LIMITING(a.lim_const), JITTER(a.jitter),
+      P_ORDERS(a.order), TARGET_IDS(a.target_id), METRIC_IDS(a.metric_id),
+      Q_ORDERS(a.quad_order), LINEAR_SOLVERS(a.lin_solver),
+      NEWTON_LOOPS(a.newton_loop)
+   { }
+
+   void Run(const int id =0) const
+   {
+      static const bool all = getenv("MFEM_TESTS_UNIT_TMOP_ALL");
+      if ((id==0) && name) { mfem::out << "[" << name << "]" << std::endl; }
+      DEFAULT_ARGS;
+      char ni[8] {}, rs[8] {}, li[8] {}, lc[16] {}, ji[16] {}, cmb[8] {};
+      args[MSH] = mesh;
+      args[RS] = itoa(REFINE,rs);
+      args[NI] = itoa(NEWTON_ITERATIONS,ni);
+      args[LI] = itoa(LINEAR_ITERATIONS,li);
+      args[CMB] = itoa(COMBO,cmb);
+      args[LC] = dtoa(LIMITING,lc);
+      args[JI] = dtoa(JITTER,ji);
+      args[NOR] = NORMALIZATION ? "1" : "0";
+      for (int p : P_ORDERS)
+      {
+         char por[2] {};
+         args[POR] = itoa(p, por);
+         for (int t : TARGET_IDS)
+         {
+            char tid[2] {};
+            args[TID] = itoa(t, tid);
+            for (int m : METRIC_IDS)
+            {
+               char mid[4] {};
+               args[MID] = itoa(m, mid);
+               for (int q : Q_ORDERS)
+               {
+                  if (q <= p) { continue; }
+                  char qor[2] {};
+                  args[QOR] = itoa(q, qor);
+                  for (int ls : LINEAR_SOLVERS)
+                  {
+                     char lsb[2] {};
+                     args[LS] = itoa(ls, lsb);
+                     for (int n : NEWTON_LOOPS)
+                     {
+                        char nl[2] {};
+                        args[NL] = itoa(n, nl);
+                        tmop_require(id, args);
+                        if (!all) { break; }
+                     }
+                     if (!all) { break; }
+                  }
+                  if (!all) { break; }
+               }
+               if (!all) { break; }
+            }
+            if (!all) { break; }
+         }
+         if (!all) { break; }
+      }
+   }
+};
+
+static void tmop_tests(int id)
```

Comment:
This function includes many runs that probably belong in `miniapps/meshing/{p,}mesh-optimizer.cpp`. In the interest of keeping the unit tests shorter, I think we should move most of these as sample runs in the mesh optimizer miniapp.

Reply:
We can keep them here commented out too -- they may be convenient for future testing of new TMOP developments.

Reply:
In fact, most of these tests are not launch by default, unless told to by setting `MFEM_TESTS_UNIT_TMOP_ALL`.
I'll try to see if we can pass it through the command line.

Reply:
With the new `--all` option you can launch even more tests, however, this did not change anything when the option is not given. So my original concern still remains.

Reply:
I've shortened the tmop tests w/o the `--all` options: it seems to be now less than 5s in the worst case instead of 20s.

Reply:
Most of these unit tests have been taken from the `miniapps/meshing/{p,}mesh-optimizer.cpp`.
There are here new combinations with some `limiting` and `jitter` hardcoded values, but I'm not sure we want to add all of them - which also have been tightened in order to be able to truly compare `fa` and `pa` runs.
There is already a lot of more numerical sample runs in  `miniapps/meshing`:
```
//  Preconditioner run:
// ./mesh-optimizer -m blade.mesh -o 4 -rs 0 -mid 2 -tid 1 -ni 200 -bnd -qt 1 -qo 8 -ls 3 -nor -lc 0.1 -vl 2 -pa -d cpu
//
// Sample runs:
//   Adapted analytic shape:
//     mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 2 -tid 4 -ni 200 -bnd -qt 1 -qo 8
//   Adapted analytic size+orientation:
//     mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 14 -tid 4 -ni 100 -bnd -qt 1 -qo 8 -fd
//   Adapted analytic shape+orientation:
//     mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 85 -tid 4 -ni 100 -bnd -qt 1 -qo 8 -fd
//
//   Adapted discrete size:
//     mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 80 -tid 5 -ni 50 -qo 4 -nor
//   Adapted discrete size; explicit combo of metrics; mixed tri/quad mesh:
//     mesh-optimizer -m ../../data/square-mixed.mesh -o 2 -rs 2 -mid 2 -tid 5 -ni 200 -bnd -qo 6 -cmb 2 -nor
//   Adapted discrete size+aspect_ratio:
//     mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 7 -tid 6 -ni 100
//     mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 7 -tid 6 -ni 100 -qo 6 -ex -st 1 -nor
//   Adapted discrete size+orientation (requires GSLIB):
//   * mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 36 -tid 8 -qo 4 -fd -ae 1 -nor
//   Adapted discrete aspect-ratio+orientation (requires GSLIB):
//   * mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 85 -tid 8 -ni 10 -bnd -qt 1 -qo 8 -fd -ae 1
//   Adapted discrete aspect ratio (3D):
//     mesh-optimizer -m cube.mesh -o 2 -rs 2 -mid 302 -tid 7 -ni 20 -bnd -qt 1 -qo 8
//
//   Adapted discrete size (3D):
//     mesh-optimizer -m cube.mesh -o 2 -rs 2 -mid 321 -tid 5 -ni 20  -ls 2 -li 100 -bnd -qt 1 -qo 8 -nor
//   Adapted discrete shape+size explicit combo (3D):
//     mesh-optimizer -m cube.mesh -o 2 -rs 2 -mid 302 -tid 5 -ni 20  -ls 2 -li 100 -bnd -qt 1 -qo 8 -cmb 2 -nor
//
//   Adaptive limiting:
//     mesh-optimizer -m stretched2D.mesh -o 2 -mid 2 -tid 1 -ni 50 -qo 5 -nor -vl 1 -alc 0.5
//   Adaptive limiting through the L-BFGS solver:
//     mesh-optimizer -m stretched2D.mesh -o 2 -mid 2 -tid 1 -ni 400 -qo 5 -nor -vl 1 -alc 0.5 -st 1
//   Adaptive limiting through FD (requires GSLIB):
//   * mesh-optimizer -m stretched2D.mesh -o 2 -mid 2 -tid 1 -ni 50 -qo 5 -nor -vl 1 -alc 0.5 -fd -ae 1
//
//   Blade shape:
//     mesh-optimizer -m blade.mesh -o 4 -mid 2 -tid 1 -ni 30 -ls 3 -art 1 -bnd -qt 1 -qo 8
//   Blade shape with FD-based solver:
//     mesh-optimizer -m blade.mesh -o 4 -mid 2 -tid 1 -ni 30 -ls 4 -bnd -qt 1 -qo 8 -fd
//   Blade limited shape:
//     mesh-optimizer -m blade.mesh -o 4 -mid 2 -tid 1 -bnd -qt 1 -qo 8 -lc 5000
//   ICF shape and equal size:
//     mesh-optimizer -o 3 -mid 9 -tid 2 -ni 25 -ls 3 -art 2 -qo 5
//   ICF shape and initial size:
//     mesh-optimizer -o 3 -mid 9 -tid 3 -ni 30 -ls 3 -bnd -qt 1 -qo 8
//   ICF shape:
//     mesh-optimizer -o 3 -mid 1 -tid 1 -ni 100 -bnd -qt 1 -qo 8
//   ICF limited shape:
//     mesh-optimizer -o 3 -mid 1 -tid 1 -ni 100 -bnd -qt 1 -qo 8 -lc 10
//   ICF combo shape + size (rings, slow convergence):
//     mesh-optimizer -o 3 -mid 1 -tid 1 -ni 1000 -bnd -qt 1 -qo 8 -cmb 1
//   Mixed tet / cube / hex mesh with limiting:
//     mesh-optimizer -m ../../data/fichera-mixed-p2.mesh -o 4 -rs 1 -mid 301 -tid 1 -fix-bnd -qo 6 -nor -lc 0.25
//   3D pinched sphere shape (the mesh is in the mfem/data GitHub repository):
//   * mesh-optimizer -m ../../../mfem_data/ball-pert.mesh -o 4 -mid 303 -tid 1 -ni 20 -li 500 -fix-bnd
//   2D non-conforming shape and equal size:
//     mesh-optimizer -m ./amr-quad-q2.mesh -o 2 -rs 1 -mid 9 -tid 2 -ni 200 -bnd -qt 1 -qo 8
//
//   2D untangling:
//     mesh-optimizer -m jagged.mesh -o 2 -mid 22 -tid 1 -ni 50 -li 50 -qo 4 -fd -vl 1
//   3D untangling (the mesh is in the mfem/data GitHub repository):
//   * mesh-optimizer -m ../../../mfem_data/cube-holes-inv.mesh -o 3 -mid 313 -tid 1 -rtol 1e-5 -li 50 -qo 4 -fd -vl 1
```